storage/innobase/trx/trx0rseg.cc

```diff
@@ -571,12 +601,48 @@ trx_rseg_array_init()
 				ut_ad(!trx_sys.rseg_array[rseg_id]);
 				trx_sys.rseg_array[rseg_id] = rseg;
 				trx_rseg_mem_restore(rseg, max_trx_id, &mtr);
+#ifdef WITH_WSREP
+				if (!wsrep_sys_xid.is_null() &&
+				    !wsrep_sys_xid.eq(&trx_sys.recovered_wsrep_xid)) {
+					wsrep_xid_in_rseg_found = true;
+					ut_ad(memcmp(wsrep_xid_uuid(&wsrep_sys_xid),
+						     wsrep_xid_uuid(&trx_sys.recovered_wsrep_xid),
+						     sizeof wsrep_uuid) ||
+					      (wsrep_xid_seqno(&wsrep_sys_xid) <=
+					       wsrep_xid_seqno(&trx_sys.recovered_wsrep_xid)));
+				}
+#endif
 			}
 		}
 
 		mtr.commit();
 	}
 
+#ifdef WITH_WSREP
+	if (!wsrep_sys_xid.is_null()) {
+		/* Upgrade from a version prior to 10.3.5,
+		where WSREP XID was stored in TRX_SYS page.
+		If no rollback segment has a WSREP XID set,
+		we must copy the XID found in TRX_SYS page
+		to rollback segments. */
+		mtr_t mtr;
+		mtr.start();
```

Comment:
This looks good. The `mtr.commit()` in the preceding loop was read-only. What matters is that the writes will be executed in a single atomic mini-transaction.